### DISCUSSION TITLE
Maybe HAL problem with ESP32-LyraT

### DISCUSSION THREAD

**Discussion by jonrichings** [pschatzmann/arduino-audiokit#15] [General]
I'm trying to setup Arduino IDE, AudioKit and AudioTools for a LyraT V4.3.  

I can compile successfully, upload and run some basic examples.  However it looks to me as though the touch sensors are not being read correctly (repetitive random activations and non-activations of touched buttons).  It also seems that the headphone jack switch is not being acted on (but speakers work).

I did the AudioKitSettings mod (#define AUDIOKIT_BOARD 1).  Maybe there is more I need to do?

**Comment by pschatzmann** [comment]
You might need to adjust the limit. I think it is something like:
```
kit.audioActions().setTouchLimit(limit);
```
Or you can just comment out   ```kit.processActions();``` to deactivate my implementation

**Reply by jonrichings** [reply]
Commenting out kit.processActions(); did prevent the awful result but I don't know how else to trigger sounds (don't have midi setup yet).

kit.audioActions().setTouchLimit(limit); I tried limit values of 1 and 100 and 254 but there was no apparent change in the touch pad behavior.  Any other ideas?

**Comment by pschatzmann** [comment]
I think if you set the log level to debug you should see the values that you receive...

**Reply by jonrichings** [reply]
It scrolls so fast it's hard to trap the input.  But this is mostly what I see:

```
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioStreams.h : 283 - GeneratedSoundStream::readBytes: 1024
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 0 -> 0 bytes - in 0 hops
[I] SoundGenerator.h : 207 - setFrequency: 196
[I] SoundGenerator.h : 208 - active: false
[D] AudioStreams.h : 249 - bool audio_tools::GeneratedSoundStream<T>::begin() [with T = short int]
[I] SoundGenerator.h : 166 - bool audio_tools::SineWaveGenerator<T>::begin() [with T = short int]
[D] SoundGenerator.h : 35 - bool audio_tools::SoundGenerator<T>::begin() [with T = short int]
[I] SoundGenerator.h : 38 - bits_per_sample: 16
[I] SoundGenerator.h : 39 - channels: 2
[I] SoundGenerator.h : 40 - sample_rate: 44100
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioStreams.h : 283 - GeneratedSoundStream::readBytes: 1024
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
```

I can only see the setFrequency: 196 part as a response to pressing a key.  I think the worst cause of the noise is multiple triggers but without key-up events to end them. I think the first two physical buttons are working but the touch sensors are just all firing like crazy and the end result is a lot of very distorted noise with barely a coherent tone.

That and the headphone socket switch not seeming to work suggests to me a hardware config issue.

**Comment by pschatzmann** [comment]
What sketch are you using ? If the keys are defined in the sketch you might need to indicate that you have touch buttons
  kit.audioActions().add(PIN_KEY1, actionKeyOn, actionKeyOff, ActiveTouch, &(note[0])); // C3

**Reply by jonrichings** [reply]
I'm using streams-synthbasic1-audiokit

Tried kit.audioActions().add(PIN_KEY1, actionKeyOn, actionKeyOff, ActiveTouch, &(note[0])); // C3

Got 'ActiveTouch' was not declared in this scope

Does that need a lib I havn't included?

**Comment by jonrichings** [comment]
Where are the PIN_KEYs defined?

**Comment by pschatzmann** [comment]
ActiveTouch is part of [AudioActions::ActiveLogic](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_audio_actions.html):
 kit.audioActions().add(PIN_KEY1, actionKeyOn, actionKeyOff, AudioActions::ActiveTouch, &(note[0])); // C3

The keys themselfs are defined in 
https://github.com/pschatzmann/arduino-audiokit/blob/main/src/audio_board/lyrat_v4_3/board_def.h

By the way here is the link to the [Class Documentation](https://pschatzmann.github.io/arduino-audio-tools/html/namespaceaudio__tools.html) that you can also find in the Readme...

ps. don't forget to change the log level to Warning to get perfect audio

**Comment by jonrichings** [comment]
Hi Phil,

The AudioActions doc didn't explain ActiveTouch but I tried a few variations with no success.  Went back to the original and it sort of works.... sometimes.   I noticed that board_def.h has` #include "driver/touch_pad.h"` commented out, so I tried uncommenting that but it made no difference.  So, I'm giving up on the buttons for now and will try to get MIDI to work as in input.

For future readers, Phil's suggestion to change the log level to Warning does make a BIG difference o audio quality.

**Comment by jonrichings** [comment]
OK, I have BLE MIDI working.  At least I can send MIDI messages from a simple phone app and see them being received correctly (Serial.print...).  But I am still getting wacky sounds from the LyraT Board.  Low notes just crackly and only a few high notes play.  I've attached my code.  I just crudely cut and pasted a couple of examples together.  I'm sure there are more elegant ways to do this but since this half works I'm confused.

``/**
   @file streams-synthbasic-audiokit.ino
   @author Phil Schatzmann
   @brief see https://www.pschatzmann.ch/home/2021/12/17/ai-thinker-audio-kit-building-a-simple-synthesizer-with-the-audiotools-library/)
   @copyright GPLv3

*/
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

#include <Midi.h>


MidiCallbackAction action;
MidiBleServer ble("MidiServer", &action);


AudioKitStream kit;
SineWaveGenerator<int16_t> sine;
GeneratedSoundStream<int16_t> in(sine);
StreamCopy copier(kit, in);

void actionKeyOn(bool active, int pin, void* ptr) {
  int freq = *((int*)ptr);
  sine.setFrequency(freq);
  in.begin();
}

void actionKeyOff(bool active, int pin, void* ptr) {
  in.end();
}

// We want to play some notes on the AudioKit keys
void setupActions() {
  // assign buttons to notes
  auto act_low = AudioActions::ActiveLow;
  //auto act_low = AudioActions::ActiveTouch;
  static int note[] = {N_C3, N_D3, N_E3, N_F3, N_G3, N_A3}; // frequencies


  kit.audioActions().add(PIN_KEY1, actionKeyOn, actionKeyOff, act_low, &(note[0])); // C3
  kit.audioActions().add(PIN_KEY2, actionKeyOn, actionKeyOff, act_low, &(note[1])); // D3
  kit.audioActions().add(PIN_KEY3, actionKeyOn, actionKeyOff, act_low, &(note[2])); // E3
  kit.audioActions().add(PIN_KEY4, actionKeyOn, actionKeyOff, act_low, &(note[3])); // F3
  kit.audioActions().add(PIN_KEY5, actionKeyOn, actionKeyOff, act_low, &(note[4])); // G3
  kit.audioActions().add(PIN_KEY6, actionKeyOn, actionKeyOff, act_low, &(note[5])); // A3

  //kit.audioActions().setTouchLimit(1);
}

void onNoteOn(uint8_t channel, uint8_t note, uint8_t velocity) {
  int freq = MidiCommon::noteToFrequency(note);
  Serial.print("onNoteOn: ");
  Serial.print(note);
  Serial.print(" ");
  Serial.println(freq);
  //int freq = *((int*)ptr);
  sine.setFrequency(freq);
  in.begin();
}

void onNoteOff(uint8_t channel, uint8_t note, uint8_t velocity) {
  Serial.print("onNoteOff: ");
  Serial.println(note);
  in.end();
}

void setup() {
  Serial.begin(115200);
  //AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);
  kit.setVolume(60);

  // Setup sound generation based on AudioKit settins
  in.begin(cfg);

  // activate keys
  setupActions();

// BLE MIDI

  action.setCallbacks(onNoteOn, onNoteOff);
  ble.begin(action);

}

// copy the data
void loop() {
  copier.copy();
  kit.processActions();
}

**Comment by pschatzmann** [comment]
If you look at your LyraT board, you can see that the first 2 buttons are regular ones and the play, sel, vol- and vol+ are Touch buttons.
They work differently. The regular buttons need to be pulled high and interpreted with digitalRead(): a low value is button pressed (hence ActiveLow), while the touch buttons need to be read with touchRead() and are active if they are higher then a defined limit. So I would expect that for your board you will need to use:

kit.audioActions().add(PIN_KEY1, actionKeyOn, actionKeyOff, AudioActions::ActiveLow;, &(note[0])); // C3
kit.audioActions().add(PIN_KEY2, actionKeyOn, actionKeyOff, AudioActions::ActiveLow;, &(note[1])); // D3
kit.audioActions().add(PIN_KEY3, actionKeyOn, actionKeyOff, AudioActions::ActiveTouch, &(note[2])); // E3
kit.audioActions().add(PIN_KEY4, actionKeyOn, actionKeyOff, AudioActions::ActiveTouch, &(note[3])); // F3
kit.audioActions().add(PIN_KEY5, actionKeyOn, actionKeyOff, AudioActions::ActiveTouch, &(note[4])); // G3
kit.audioActions().add(PIN_KEY6, actionKeyOn, actionKeyOff, AudioActions::ActiveTouch, &(note[5])); // A3

I assume that your midi is not working because of the enterfering buttons: just comment out the // kit.processActions(); to make sure that this is not the case

**Reply by jonrichings** [reply]
The audioActions are what I had (without the rogue ; in the top two lines).  The two buttons work (albeit with a crackle on keyoff), nothing from the touch sensors.  I've tried changing kit.audioActions().setTouchLimit(10); from 1 to 255).  BTW, running just the touch sensor example which outputs the touch values on serial works just fine.

Still the same garbled sounds with // kit.processActions();  It is odd.  I am using an app that can play a virtual keyboard or send complete tunes as a midi sequence.  While the code is obviously responding it seems only to be able to play 3 or four notes that are roughly corresponding to the sent sequence (like high, medium, low frequencies rather than their exact note).  Most of the lower key don't produce a note at all, just a crackle.
